src/core/algorithms/fd/eulerfd/eulerfd.cpp

```diff
@@ -0,0 +1,427 @@
+#include "eulerfd.h"
+
+namespace algos {
+
+EulerFD::EulerFD() : FDAlgorithm({kDefaultPhaseName}), mlfq_(kQueuesNumber) {
+    last_ncover_ratios_.fill(1);
+    last_pcover_ratios_.fill(1);
+    RegisterOption(config::kCustomRandomFlagOpt(&custom_random_opt_));
+
+    // set configuration options
+    RegisterOption(config::kTableOpt(&input_table_));
+    RegisterOption(config::kEqualNullsOpt(&is_null_equal_null_));
+    MakeOptionsAvailable({config::kTableOpt.GetName(), config::kEqualNullsOpt.GetName()});
+
+    max_lhs_ = std::numeric_limits<unsigned int>::max();
+}
+
+void EulerFD::MakeExecuteOptsAvailable() {
+    MakeOptionsAvailable({config::kCustomRandomFlagOpt.GetName()});
+}
+
+void EulerFD::LoadDataInternal() {
+    number_of_attributes_ = input_table_->GetNumberOfColumns();
+    if (number_of_attributes_ == 0) {
+        throw std::runtime_error("Unable to work on an empty dataset.");
+    }
+
+    schema_ = std::make_unique<RelationalSchema>(input_table_->GetRelationName());
+    for (size_t i = 0; i < number_of_attributes_; ++i) {
+        std::string const &column_name = input_table_->GetColumnName(static_cast<int>(i));
+        schema_->AppendColumn(column_name);
+    }
+
+    // in each column mapping string values into integer values.
+    // using only hash isnt good idea because colisions dont processing
+    std::vector<std::unordered_map<std::string, size_t>> columns(number_of_attributes_);
```

Comment:
There is no unnecessary copy, because all maps "indexes" (strings) was moved (`values[std::move(line[i])] = id`, line 54) from std::vector<std::string>, on which allocation I can't influence, because it is a result of `input_table_->GetNextRow()` method.